### PR TITLE
Result based constructor

### PR DESCRIPTION
**Pull request by faern**
Since the inputs can be invalid I thought it would be reasonable to have checks for that. What do you think? Implemented with a special error enum type.

### CODE REVIEW

**Comment by faern** [comment]
Also changed all previous `Result` returns to use the new enum instead of just `String` for error type.

File: src/lib.rs [hunk 1]
```diff
@@ -275,6 +298,12 @@ mod test {
     }
 
     #[test]
+    fn parse_v4_non_zero_host_bits() {
+        let cidr = Ipv4Network::from_cidr("10.1.1.1/24");
+        assert!(cidr.is_err());
```
Comment (achanda) [inline src/lib.rs]: I don't think this should be rejected. This shows an example http://mxtoolbox.com/subnet/?filter=10.1.1.0/24&source=subnettool

**Comment by achanda** [comment]
Returning a `Result` makes perfect sense. One small comment though.

File: src/lib.rs (hunk 1 above)
Comment (faern) [inline src/lib.rs]: I think I missed the point you tried to make with that link. I can't find a CIDR with non-zero host bits in that link. However I found that Apache Commons accepts it so I agree with your anyway. https://commons.apache.org/proper/commons-net/apidocs/org/apache/commons/net/util/SubnetUtils.html

**Pull request merged**